service/history/queues/action_pending_task_count.go

```diff
@@ -0,0 +1,226 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package queues
+
+import (
+	"time"
+
+	"golang.org/x/exp/slices"
+
+	"go.temporal.io/server/common/collection"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+var (
+	targetThresholdCoefficient = 0.8
+	clearSliceThrottleDuration = 10 * time.Second
+)
+
+type (
+	actionQueuePendingTask struct {
+		attributes     *AlertAttributesQueuePendingTaskCount
+		monitor        Monitor
+		maxReaderCount int
+		completionFn   actionCompletionFn
+
+		// state of the action, used when running the action
+		tasksPerNamespace               map[namespace.ID]int
+		pendingTaskPerNamespacePerSlice map[Slice]map[namespace.ID]int
+		slicesPerNamespace              map[namespace.ID][]Slice
+		namespaceToClearPerSlice        map[Slice][]namespace.ID
+	}
+)
+
+func newQueuePendingTaskAction(
+	attributes *AlertAttributesQueuePendingTaskCount,
+	monitor Monitor,
+	maxReaderCount int,
+	completionFn actionCompletionFn,
+) Action {
+	return &actionQueuePendingTask{
+		attributes:     attributes,
+		monitor:        monitor,
+		maxReaderCount: maxReaderCount,
+		completionFn:   completionFn,
+	}
+}
+
+func (a *actionQueuePendingTask) Run(readerGroup *ReaderGroup) {
```

Comment:
I will send separate pr for adding unit test for all action implementations. Will test the impl locally to verify it works. Sending out the impl PR first to speed up the review process.